File: 2022/day-6/day6.py

def find_sequence(inp: str, message=4) -> int:
    for i in range(0, len(inp)-message):
        subset = inp[i:i+message]
        if len(set(subset)) == message:
            return i + message


with open('input.txt', 'r') as h:
    m = h.read()

print(find_sequence(m))
print(find_sequence(m, 14))

File: 2022/day-6/day6.rs
extern crate core;

use std::collections::HashSet;

fn find_message_start(message: &str, length: usize) -> Option<usize> {
    for i in 0usize..(message.len() - length) {
        let subset = &message[i..i + length];
        if subset.chars().collect::<HashSet<char>>().len() == length {
            return Some(i + length);
        }
    }
    None
}

fn main() {
    let contents = std::fs::read_to_string("input.txt").expect("Failed to read input");
    println!("Part 1: {}", find_message_start(&contents, 4).unwrap());
    println!("Part 2: {}", find_message_start(&contents, 14).unwrap());
}
